feat(docs): add cross chain mint and redeem guides (#1377)
flare-foundation/developer-hub: docs/smart-accounts/guides/typescript-viem/05-cross-chain-redeem.mdx

---

> sidebar_position: 1

Reviewer: Fix sidebar position

slug: cross-chain-redeem-ts
title: Cross-Chain Redeem
authors: [nikerzetic]
description: Send FXRP from Sepolia back to Flare and auto-redeem it to XRP on the XRPL using Viem and LayerZero.
tags: [intermediate, ethereum, flare-smart-accounts]
keywords:
  [
    flare-fdc,
    ethereum,
    flare-smart-accounts,
    evm,
    flare-network,
    account-abstraction,
    layer-zero,
    fxrp,
  ]
unlisted: false
---

import CodeBlock from "@theme/CodeBlock";
import CrossChainRedeemScript from "!!raw-loader!/examples/developer-hub-javascript/smart-accounts/cross-chain-redeem.ts";

In this guide, you will learn how to send FXRP from Sepolia back to Flare and trigger an automatic redemption to native XRP on the XRPL, all in a single LayerZero transaction.
The flow is:

- the user holds FXRP on Sepolia and calls `send` on the FXRP OFT with a compose message attached
- the message is delivered to a composer contract on Flare, which calls the `AssetManager` to redeem the FXRP for XRP

> - the XRP is paid out by an agent to the XRPL address specified in the compose message — without a destination tag

Reviewer: ```suggestion
In this guide, you will learn how to send FXRP from Sepolia back to Flare and trigger an automatic redemption to native XRP on the XRPL, all in a single LayerZero transaction.

The flow is:
- the user holds FXRP on Sepolia and calls `send` on the FXRP OFT with a compose message attached.
- the message is delivered to a composer contract on Flare, which calls the `AssetManager` to redeem the FXRP for XRP.
- the XRP is paid out by an agent to the XRPL address specified in the compose message — without a destination tag.
```

> A prerequisite for the script to work is a Sepolia EOA that already holds FXRP and enough SETH to cover the LayerZero native fee.

Reviewer: Write in full words `EOA` first time

If you need to obtain FXRP on Sepolia first, see the [Cross-Chain Mint guide](/smart-accounts/guides/typescript-viem/cross-chain-mint-ts).
The default composer address (`0xa10569DFb38FE7Be211aCe4E4A566Cea387023b0`) can be overridden via the `COSTON2_COMPOSER` environment variable; the amount and destination are controlled by `SEND_LOTS` and `XRP_ADDRESS`.

The full code showcased in this guide is available on [GitHub](https://github.com/flare-foundation/flare-viem-starter).

:::info

> The code in this guide is set up for the Coston2 testnet and the Sepolia testnet.

Reviewer: ```suggestion
The code in this guide is set up for the Coston2 and Sepolia testnets.
```

```

```typescript title="src/cross-chain-redeem.ts"
const { nativeFee, lzTokenFee } = await sepoliaPublicClient.readContract({
  address: oftAddress,
  abi: fxrpOftAbi,
  functionName: "quoteSend",
  args: [sendParam, false],
});
```

The fee is denominated in ETH because that is the chain the transaction originates on.
The `quoteSend` function accounts for the compose payload — a larger `composeMsg` increases the quoted fee.

## Sending the transaction

The send is performed with the standard Viem `simulateContract` + `writeContract` pattern, attaching the quoted native fee as the transaction value.

```typescript title="src/cross-chain-redeem.ts"
const { request } = await sepoliaPublicClient.simulateContract({
  account,
  address: oftAddress,
  abi: fxrpOftAbi,
  functionName: "send",
  args: [sendParam, { nativeFee, lzTokenFee }, signerAddress],
  value: nativeFee,
});

const txHash = await sepoliaWalletClient.writeContract(request);
```

The third argument of `send` is the refund address — any unused portion of `nativeFee` is returned there.
Once the Sepolia receipt is in, the LayerZero scanner link is printed so the user can follow the cross-chain delivery in real time.

> ## Waiting for `RedemptionRequested`

Reviewer: ```suggestion
## Waiting for `RedemptionRequested` Event
```

> A successful flow results in the `AssetManager` on Flare emitting a `RedemptionRequested` event with the composer as the `redeemer` indexed argument.

Reviewer: Link to `RedemptionRequested` event in reference

  const deadline = Date.now() + REDEMPTION_TIMEOUT_MS;

  let cursor = fromBlock;
  while (Date.now() < deadline) {
    const latest = await publicClient.getBlockNumber();
    while (cursor <= latest) {
      const chunkEnd = cursor + MAX_BLOCK_RANGE - 1n;
      const toBlock = chunkEnd > latest ? latest : chunkEnd;
      const logs = await publicClient.getContractEvents({
        address: assetManagerAddress,
        abi: coston2.iAssetManagerAbi,
        eventName: "RedemptionRequested",
        args: { redeemer: CONFIG.COSTON2_COMPOSER },
        fromBlock: cursor,
        toBlock,
      });
      if (logs.length > 0) {
        return logs[0]!;
      }
      cursor = toBlock + 1n;
    }
    await new Promise((resolve) =>
      setTimeout(resolve, REDEMPTION_POLL_INTERVAL_MS),
    );
  }

  throw new Error(
    `RedemptionRequested event not observed within ${REDEMPTION_TIMEOUT_MS}ms`,
  );
}
```

The inner loop advances `cursor` chunk-by-chunk through any new blocks that have appeared, then sleeps 10 seconds before re-checking the chain tip.
Filtering by `redeemer` on the indexed argument means the call only returns events tied to the composer this script is targeting.

> The actual XRP payment to `underlyingAddress` is handled off-chain by an F-Asset agent after the redemption request is filed.

Reviewer: ```suggestion
The actual XRP payment to `underlyingAddress` is handled off-chain by an FAsset agent after the redemption request is filed.
```

That step is asynchronous and is not awaited by this script.

## Full script

The repository with the above example is available on [GitHub](https://github.com/flare-foundation/flare-viem-starter).
In the example repository, certain helpers are isolated into separate files in the `src/utils` directory, and the shared `SendParam` type lives in `src/layer-zero/types.ts`.

<CodeBlock language="typescript" title="src/cross-chain-redeem.ts">
  {CrossChainRedeemScript}
</CodeBlock>

## Expected output

```bash

> nik ~/g/g/f/s/flare-smart-accounts-viem on main !8 via node-v24.14.1

Reviewer: Remove your user in output for safety reasons.

> pnpm run script src/layer-zero/cross-chain-redeem.ts

> > @flarenetwork/flare-smart-accounts-viem@0.0.1 script /home/nik/git/gitlab.com/flarenetwork/starters-examples/flare-smart-accounts-viem

Reviewer: Here as well remove your username in console

    '0x0000000000000000000000000000000000000000000000000000000001bd8652'
  ],
  data: '0x00000000000000000000000000000000000000000000000000000000000001200000000000000000000000000000000000000000000000000000000000989680000000000000000000000000000000000000000000000000000000000000c350000000000000000000000000000000000000000000000000000000000108148500000000000000000000000000000000000000000000000000000000010816f0000000000000000000000000000000000000000000000000000000006a02f8194642505266410002000000000000000000000000000000000000000001bd8652000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000022727048757734624b536a6f6e4b52724b4b5659555a595956656467316a7950726d70000000000000000000000000000000000000000000000000000000000000',
  blockNumber: 30438649n,
  transactionHash: '0x9d82fe98f04fa37f729c786cab99632c781b9f3bc655cdd040a8adcfc497f736',
  transactionIndex: 1,
  blockHash: '0xa006cf7988226e5e57a1194e9c5a183beec2a37d999f966199f943fe044a77a2',
  logIndex: 9,
  removed: false,
  blockTimestamp: undefined
}

> ```

Reviewer: Add what's next block with

* mint
* redeem to tag
* direct minting overview
